Addons/tsgExoticQuadrature.hpp

```diff
@@ -97,23 +107,29 @@ inline double lagrange_eval(size_t idx, const std::vector<double> &roots, double
     return eval_value;
 }
 
-//! \brief Generate the full cache of Exotic (weighted and possibly shifted) Gauss-Legendre weights and points for
-//! an (input) max level \b n, using parameters \b shift, \b weight_fn, and \b nref. By default, the reference points
-//! and reference weights are generated by the Gauss-Legendre rule.
+//! \brief Generate the full cache of Exotic (weighted and possibly shifted) quadrature weights and points for an (input) max level
+//! \b n, using parameters \b shift, \b weight_vals, \b ref_points, and \b nref.
 //!
-//! The integrand is assumed to be of the form F(x) = f(x) * weight_fn(x) and the function x->weight_fn(x)+shift
-//! is nonnegative.
-inline TasGrid::CustomTabulated getExoticQuadrature(const int n, const double shift, std::function<double(double)> weight_fn,
-                                                    const int nref, const char* description) {
+//! The integral  I := ∫F(x)*weight_fn(x)dx, for a given level, is approximated as:
+//! I ≈ sum{f(roots[i])*(weight_fn_vals[i]+shift)*ref_weights[i]} - shift * sum{f(correction_points[i])*correction_weights[i]}.
+//!
+//! It is assumed that (weight_fn_vals[i]+shift) is nonnegative for every i.
+inline TasGrid::CustomTabulated getExoticQuadrature(const int n, const double shift, const std::vector<double> &weight_fn_vals,
+                                                    std::vector<double> &ref_weights, const std::vector<double> &ref_points,
```

Comment:
Notice how everything `const` except `ref_weights` which gets modified, but it is not an output.

Quirks like this can waste a lot of time if someone is trying to figure out what's happening here. It may even be you and me couple of years from now.

Think of a way to solve this and we can discuss it in person.

Reply:
My first thought about this would be to make ``ref_weights`` a ``const`` input and just create a new array inside the function that stores the entries ``ref_weights[i] * (weight_fn_vals[i]  + shift)``, but this might not be the most efficient way. 

On the other hand, I'm not sure if I _can_ save on storage when all of the inputs are ``const`` lvalues at the moment.

Reply:
Your first thought was better than the current solution, since the interface is simpler and more consistent and the extra copy is a very cheap price to pay.

But there is an even better way. Why not pass in the modified `ref_weights` directly. Think of how that would work.

Reply:
Gotcha. I could rename ``ref_weights`` to something like ``shifted_weights`` and have it already contain the values ``ref_weights[i] * (weight_fn_vals[i] + shift)``. Then, I would modify the overloaded function (which takes in a lambda) appropriately. 

I was slightly against going this direction because the user has to be very careful in making sure that this variable is generated properly, and there is no easy way to check to see if they made a mistake.

Reply:
> I was slightly against going this direction because the user has to be very careful in making sure that this variable is generated properly, and there is no easy way to check to see if they made a mistake.

Your observation is spot on and this is a real issue. Think of how we can resolve that.

Reply:
Changed to use ``shifted_weights`` for now. Will make an external/internal API in another pull request based on our earlier in-person conversation.